src/main/java/com/google/devtools/build/lib/rules/cpp/Cpp20ModuleDepMapAction.java

```diff
@@ -0,0 +1,222 @@
+// Copyright 2014 The Bazel Authors. All rights reserved.
```

Comment:
Do you mean write starlark in builtins_bzl, i.e `main/starlark/builtins_bzl/common/cc/cc_binary.bzl`?

Reply:
I think if you try doing it in the `builtins_bzl` it would become more clear what I mean.

I think this whole action can be constructed in Starlark by calling: `ctx.actions.run`. The `execute` method seems complex. But imagine you have a shell tools, a cc_binary, that accepts arguments with input files and does exactly what `execute` method does, writing given output files. Then you can use that in `ctx.actions.run(tools = ...)`. 

The code for the tool can go into `/tools/cc`.

You need to bring in an implicit dependency that points to such a tool. You can do that on `cc_toolchain`. Follow the example of `_grep_include`.

If you can do the `ctx.actions.run` in `builtins_bzl`, that's perfect. 

If you can't, then you can still natively construct a SpawnAction with everything that `ctx.actions.run` does (but without creating a dedicated class for it.

Reply:
Two tools, `generate-modmap` and `aggregate-ddi`, have been introduced to replace the Java class implementation. These tools are utilized via `SpawnAction`.

Currently, it is challenging to directly use `generate-modmap` and `aggregate-ddi` in `builtins_bzl`. This is because the logic for constructing the action graph currently resides in `cc_common.compile`, which is implemented in Java and has not yet been migrated to Starlark.

Reply:
the changes about constructing the action graph are on the patch 4
https://github.com/bazelbuild/bazel/pull/22553